File: html-generators/generateog.java
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 25
//DEPS com.fasterxml.jackson.core:jackson-databind:2.18.3
//DEPS com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:2.18.3
//DEPS org.apache.xmlgraphics:batik-transcoder:1.18
//DEPS org.apache.xmlgraphics:batik-codec:1.18

import module java.base;
import com.fasterxml.jackson.databind.*;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import org.apache.batik.transcoder.TranscoderInput;
import org.apache.batik.transcoder.TranscoderOutput;
import org.apache.batik.transcoder.image.PNGTranscoder;

/**
 * Generate Open Graph SVG cards (1200×630) for each pattern.
 * Light theme, side-by-side Old/Modern code, slug title at top.
 *
 * Usage: jbang html-generators/generate-og.java [category/slug]
 *        No arguments → generate all patterns.
 */
static final String CONTENT_DIR = "content";
static final String OUTPUT_DIR = "site/og";
static final String CATEGORIES_FILE = "html-generators/categories.properties";

static final ObjectMapper JSON_MAPPER = new ObjectMapper();
static final ObjectMapper YAML_MAPPER = new ObjectMapper(new YAMLFactory());
static final Map<String, ObjectMapper> MAPPERS = Map.of(
    "json", JSON_MAPPER, "yaml", YAML_MAPPER, "yml", YAML_MAPPER
);

static final SequencedMap<String, String> CATEGORY_DISPLAY = loadProperties(CATEGORIES_FILE);

// ── Light-theme palette ────────────────────────────────────────────────
static final String BG        = "#ffffff";
static final String BORDER    = "#d8d8e0";
static final String TEXT       = "#1a1a2e";
static final String TEXT_MUTED = "#6b7280";
static final String OLD_BG     = "#fef2f2";
static final String MODERN_BG  = "#eff6ff";
static final String OLD_ACCENT = "#dc2626";
static final String GREEN      = "#059669";
static final String ACCENT     = "#6366f1";
static final String BADGE_BG   = "#f3f4f6";

// ── Syntax highlight colors (VS Code light-inspired) ───────────────────
static final String SYN_KEYWORD    = "#7c3aed"; // purple — keywords & modifiers
static final String SYN_TYPE       = "#0e7490"; // teal   — type names
static final String SYN_STRING     = "#059669"; // green  — strings & chars
static final String SYN_COMMENT    = "#6b7280"; // gray   — comments
static final String SYN_ANNOTATION = "#b45309"; // amber  — annotations
static final String SYN_NUMBER     = "#c2410c"; // orange — numeric literals
static final String SYN_DEFAULT    = "#1a1a2e"; // dark   — everything else

static final Set<String> JAVA_KEYWORDS = Set.of(
    "abstract", "assert", "boolean", "break", "byte", "case", "catch", "char",
    "class", "const", "continue", "default", "do", "double", "else", "enum",
    "extends", "final", "finally", "float", "for", "goto", "if", "implements",
    "import", "instanceof", "int", "interface", "long", "native", "new", "null",
    "package", "private", "protected", "public", "record", "return", "sealed",
    "short", "static", "strictfp", "super", "switch", "synchronized", "this",
    "throw", "throws", "transient", "try", "var", "void", "volatile", "when",
    "while", "with", "yield", "permits", "non-sealed", "module", "open", "opens",
    "requires", "exports", "provides", "to", "uses", "transitive",
    "true", "false"
);

static final Pattern SYN_PATTERN = Pattern.compile(
    "(?<comment>//.*)|" +                           // line comment
    "(?<blockcomment>/\\*.*?\\*/)|" +               // block comment (single line)
    "(?<annotation>@\\w+)|" +                       // annotation
    "(?<string>\"\"\"[\\s\\S]*?\"\"\"|\"(?:[^\"\\\\]|\\\\.)*\"|'(?:[^'\\\\]|\\\\.)*')|" + // strings
    "(?<number>\\b\\d[\\d_.]*[dDfFlL]?\\b)|" +      // numbers
    "(?<word>\\b[A-Za-z_]\\w*\\b)|" +               // words (keywords or identifiers)
    "(?<other>[^\\s])"                              // other single chars
);

// ── Dimensions ─────────────────────────────────────────────────────────
static final int W = 1200, H = 630;
static final int PAD = 40;
static final int HEADER_H = 100;
static final int FOOTER_H = 56;
static final int CODE_TOP = HEADER_H;
static final int CODE_H = H - HEADER_H - FOOTER_H;
static final int COL_W = (W - PAD * 2 - 20) / 2;  // 20px gap between panels
static final int CODE_PAD = 14;                     // padding inside each panel
static final int LABEL_H = 32;                      // space reserved for label above code
static final int USABLE_W = COL_W - CODE_PAD * 2;  // usable width for code text
static final int USABLE_H = CODE_H - LABEL_H - CODE_PAD; // usable height for code text
static final double CHAR_WIDTH_RATIO = 0.6;         // monospace char width ≈ 0.6 × font size
static final double LINE_HEIGHT_RATIO = 1.55;       // line height ≈ 1.55 × font size
static final int MIN_CODE_FONT = 9;
static final int MAX_CODE_FONT = 16;

// ── Helpers ────────────────────────────────────────────────────────────
static final Path FONT_CACHE = Path.of(System.getProperty("user.home"), ".cache", "javaevolved-fonts");

static final Map<String, String> FONT_URLS = Map.of(
    "Inter-Regular.ttf",
        "https://fonts.gstatic.com/s/inter/v20/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuLyfMZg.ttf",
    "Inter-Medium.ttf",
        "https://fonts.gstatic.com/s/inter/v20/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuI6fMZg.ttf",
    "Inter-SemiBold.ttf",
        "https://fonts.gstatic.com/s/inter/v20/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuGKYMZg.ttf",
    "Inter-Bold.ttf",
        "https://fonts.gstatic.com/s/inter/v20/UcCO3FwrK3iLTeHuS_nVMrMxCp50SjIw2boKoduKmMEVuFuYMZg.ttf",
    "JetBrainsMono-Regular.ttf",
        "https://fonts.gstatic.com/s/jetbrainsmono/v24/tDbY2o-flEEny0FZhsfKu5WU4zr3E_BX0PnT8RD8yKxjPQ.ttf",
    "JetBrainsMono-Medium.ttf",
        "https://fonts.gstatic.com/s/jetbrainsmono/v24/tDbY2o-flEEny0FZhsfKu5WU4zr3E_BX0PnT8RD8-qxjPQ.ttf"
);

/** Download fonts to cache and register with Java's graphics environment. */
static void ensureFonts() throws IOException {
    Files.createDirectories(FONT_CACHE);
    var ge = java.awt.GraphicsEnvironment.getLocalGraphicsEnvironment();
    for (var entry : FONT_URLS.entrySet()) {
        var file = FONT_CACHE.resolve(entry.getKey());
        if (!Files.exists(file)) {
            IO.println("Downloading %s...".formatted(entry.getKey()));
            try (var in = URI.create(entry.getValue()).toURL().openStream()) {
                Files.copy(in, file);
            }
        }
        try {
            var font = java.awt.Font.createFont(java.awt.Font.TRUETYPE_FONT, file.toFile());
            ge.registerFont(font);
        } catch (java.awt.FontFormatException e) {
            IO.println("[WARN] Could not register font %s: %s".formatted(entry.getKey(), e.getMessage()));
        }
    }
}

/** Convert an SVG string to a PNG file using Batik. */
static void svgToPng(String svgContent, Path pngPath) throws Exception {
    var input = new TranscoderInput(new java.io.StringReader(svgContent));
    try (var out = new java.io.BufferedOutputStream(Files.newOutputStream(pngPath))) {
        var transcoder = new PNGTranscoder();
        transcoder.addTranscodingHint(PNGTranscoder.KEY_WIDTH, (float) W * 2);
        transcoder.addTranscodingHint(PNGTranscoder.KEY_HEIGHT, (float) H * 2);
        transcoder.transcode(input, new TranscoderOutput(out));
    }
}

static SequencedMap<String, String> loadProperties(String file) {
    try {
        var map = new LinkedHashMap<String, String>();
        for (var line : Files.readAllLines(Path.of(file))) {
            line = line.strip();
            if (line.isEmpty() || line.startsWith("#")) continue;
            var idx = line.indexOf('=');
            if (idx > 0) map.put(line.substring(0, idx).strip(), line.substring(idx + 1).strip());
        }
        return map;
    } catch (IOException e) { throw new UncheckedIOException(e); }
}

static String xmlEscape(String s) {
    return s == null ? ""
        : s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
           .replace("\"", "&quot;").replace("'", "&apos;");
}

record Snippet(JsonNode node) {
    String get(String f)        { return node.get(f).asText(); }
    String slug()               { return get("slug"); }
    String category()           { return get("category"); }
    String title()              { return get("title"); }
    String jdkVersion()         { return get("jdkVersion"); }
    String oldCode()            { return get("oldCode"); }
    String modernCode()         { return get("modernCode"); }
    String oldApproach()        { return get("oldApproach"); }
    String modernApproach()     { return get("modernApproach"); }
    String oldLabel()           { return get("oldLabel"); }
    String modernLabel()        { return get("modernLabel"); }
    String key()                { return category() + "/" + slug(); }
    String catDisplay()         { return CATEGORY_DISPLAY.get(category()); }
}

SequencedMap<String, Snippet> loadAllSnippets() throws IOException {
    var snippets = new LinkedHashMap<String, Snippet>();
    for (var cat : CATEGORY_DISPLAY.sequencedKeySet()) {
        var catDir = Path.of(CONTENT_DIR, cat);
        if (!Files.isDirectory(catDir)) continue;
        var sorted = new ArrayList<Path>();
        for (var ext : MAPPERS.keySet()) {
            try (var stream = Files.newDirectoryStream(catDir, "*." + ext)) {
                stream.forEach(sorted::add);
            }
        }
        sorted.sort(Path::compareTo);
        for (var path : sorted) {
            var ext = path.getFileName().toString();
            ext = ext.substring(ext.lastIndexOf('.') + 1);
            var snippet = new Snippet(MAPPERS.get(ext).readTree(path.toFile()));
            snippets.put(snippet.key(), snippet);
        }
    }
    return snippets;
}

// ── SVG rendering ──────────────────────────────────────────────────────

/** Compute the best font size (MIN–MAX) that fits both code blocks in their panels. */
static int bestFontSize(List<String> oldLines, List<String> modernLines) {
    int maxChars = Math.max(
        oldLines.stream().mapToInt(String::length).max().orElse(1),
        modernLines.stream().mapToInt(String::length).max().orElse(1)
    );
    int maxLines = Math.max(oldLines.size(), modernLines.size());

    // Largest font where the widest line fits the panel width
    int byWidth  = (int) (USABLE_W / (maxChars * CHAR_WIDTH_RATIO));
    // Largest font where all lines fit the panel height
    int byHeight = (int) (USABLE_H / (maxLines * LINE_HEIGHT_RATIO));

    return Math.max(MIN_CODE_FONT, Math.min(MAX_CODE_FONT, Math.min(byWidth, byHeight)));
}

/** Truncate lines to fit the panel height at the given font size. */
static List<String> fitLines(List<String> lines, int fontSize) {
    int lineH = (int) (fontSize * LINE_HEIGHT_RATIO);
    int maxLines = USABLE_H / lineH;
    if (lines.size() <= maxLines) return lines;
    var truncated = new ArrayList<>(lines.subList(0, maxLines - 1));
    truncated.add("...");
    return truncated;
}

/** Syntax-highlight a single line of Java, returning SVG tspan fragments. */
static String highlightLine(String line) {
    if (line.equals("...")) return xmlEscape(line);
    var sb = new StringBuilder();
    var m = SYN_PATTERN.matcher(line);
    int last = 0;
    while (m.find()) {
        // append any skipped whitespace
        if (m.start() > last) sb.append(xmlEscape(line.substring(last, m.start())));
        last = m.end();
        var token = m.group();
        String color = null;
        if (m.group("comment") != null || m.group("blockcomment") != null) {
            color = SYN_COMMENT;
        } else if (m.group("annotation") != null) {
            color = SYN_ANNOTATION;
        } else if (m.group("string") != null) {
            color = SYN_STRING;
        } else if (m.group("number") != null) {
            color = SYN_NUMBER;
        } else if (m.group("word") != null) {
            if (JAVA_KEYWORDS.contains(token)) {
                color = SYN_KEYWORD;
            } else if (Character.isUpperCase(token.charAt(0))) {
                color = SYN_TYPE;
            }
        }
        if (color != null) {
            sb.append("<tspan fill=\"").append(color).append("\">").append(xmlEscape(token)).append("</tspan>");
        } else {
            sb.append(xmlEscape(token));
        }
    }
    if (last < line.length()) sb.append(xmlEscape(line.substring(last)));
    return sb.toString();
}

/** Render a column of code lines as SVG <text> elements with syntax highlighting. */
static String renderCodeBlock(List<String> lines, int x, int y, int lineH) {
    var sb = new StringBuilder();
    for (int i = 0; i < lines.size(); i++) {
        sb.append("    <text x=\"%d\" y=\"%d\" class=\"code\" xml:space=\"preserve\">%s</text>\n"
            .formatted(x, y + i * lineH, highlightLine(lines.get(i))));
    }
    return sb.toString();
}

static String generateSvg(Snippet s) {
    int leftX  = PAD;
    int rightX = PAD + COL_W + 20;
    int labelY = CODE_TOP + 26;
    int codeY  = CODE_TOP + 52;

    var rawOldLines = s.oldCode().lines().toList();
    var rawModernLines = s.modernCode().lines().toList();

    int fontSize = bestFontSize(rawOldLines, rawModernLines);
    int lineH = (int) (fontSize * LINE_HEIGHT_RATIO);

    var oldLines = fitLines(rawOldLines, fontSize);
    var modernLines = fitLines(rawModernLines, fontSize);

    return """
<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="%d" height="%d" viewBox="0 0 %d %d">
  <defs>
    <style>
      .title    { font: 700 24px/1 'Inter', sans-serif; fill: %s; }
      .category { font: 600 13px/1 'Inter', sans-serif; fill: %s; }
      .label    { font: 600 11px/1 'Inter', sans-serif; text-transform: uppercase; letter-spacing: 0.05em; }
      .code     { font: 400 %dpx/1 'JetBrains Mono', monospace; fill: %s; }
      .footer   { font: 500 13px/1 'Inter', sans-serif; fill: %s; }
      .brand    { font: 700 14px/1 'Inter', sans-serif; fill: %s; }
    </style>
    <clipPath id="clip-left">
      <rect x="%d" y="%d" width="%d" height="%d" rx="8"/>
    </clipPath>
    <clipPath id="clip-right">
      <rect x="%d" y="%d" width="%d" height="%d" rx="8"/>
    </clipPath>
  </defs>

  <!-- Background -->
  <rect width="%d" height="%d" rx="16" fill="%s"/>
  <rect x="0.5" y="0.5" width="%d" height="%d" rx="16" fill="none" stroke="%s" stroke-width="1"/>

  <!-- Header: category badge + title -->
  <rect x="%d" y="%d" width="%d" height="22" rx="6" fill="%s"/>
  <text x="%d" y="%d" class="category">%s</text>
  <text x="%d" y="%d" class="title">%s</text>

  <!-- Left panel: Old code -->
  <rect x="%d" y="%d" width="%d" height="%d" rx="8" fill="%s"/>
  <rect x="%d" y="%d" width="%d" height="%d" rx="8" fill="none" stroke="%s" stroke-width="0.5"/>
  <text x="%d" y="%d" class="label" fill="%s">✗  %s</text>
  <g clip-path="url(#clip-left)">
%s  </g>

  <!-- Right panel: Modern code -->
  <rect x="%d" y="%d" width="%d" height="%d" rx="8" fill="%s"/>
  <rect x="%d" y="%d" width="%d" height="%d" rx="8" fill="none" stroke="%s" stroke-width="0.5"/>
  <text x="%d" y="%d" class="label" fill="%s">✓  %s</text>
  <g clip-path="url(#clip-right)">
%s  </g>

  <!-- Footer -->
  <text x="%d" y="%d" class="footer">JDK %s+</text>
  <text x="%d" y="%d" class="brand">javaevolved.github.io</text>
</svg>
""".formatted(
        // viewBox
        W, H, W, H,
        // style fills
        TEXT, TEXT_MUTED, fontSize, TEXT, TEXT_MUTED, ACCENT,
        // clip-left
        leftX, CODE_TOP, COL_W, CODE_H,
        // clip-right
        rightX, CODE_TOP, COL_W, CODE_H,
        // background
        W, H, BG, W - 1, H - 1, BORDER,
        // header badge
        PAD, 28, xmlEscape(s.catDisplay()).length() * 8 + 16, BADGE_BG,
        PAD + 8, 43, xmlEscape(s.catDisplay()),
        // title
        PAD, 76, xmlEscape(s.title()),
        // left panel bg + border
        leftX, CODE_TOP, COL_W, CODE_H, OLD_BG,
        leftX, CODE_TOP, COL_W, CODE_H, BORDER,
        // left label
        leftX + 14, labelY, OLD_ACCENT, xmlEscape(s.oldLabel()),
        // left code
        renderCodeBlock(oldLines, leftX + 14, codeY, lineH),
        // right panel bg + border
        rightX, CODE_TOP, COL_W, CODE_H, MODERN_BG,
        rightX, CODE_TOP, COL_W, CODE_H, BORDER,
        // right label
        rightX + 14, labelY, GREEN, xmlEscape(s.modernLabel()),
        // right code
        renderCodeBlock(modernLines, rightX + 14, codeY, lineH),
        // footer text
        PAD, H - 22, s.jdkVersion(),
        W - PAD, H - 22,
        // need text-anchor for brand — handled in the template
        ""  // unused but keeps format args aligned
    ).replace(
        // Right-align the brand text
        "class=\"brand\">javaevolved.github.io</text>",
        "class=\"brand\" text-anchor=\"end\">javaevolved.github.io</text>"
    );
}

// ── Main ───────────────────────────────────────────────────────────────
void main(String... args) throws Exception {
    ensureFonts();

    var allSnippets = loadAllSnippets();
    IO.println("Loaded %d snippets".formatted(allSnippets.size()));

    // Filter to a single slug if provided
    Collection<Snippet> targets;
    if (args.length > 0) {
        var key = args[0];
        if (!allSnippets.containsKey(key)) {
            IO.println("Unknown pattern: " + key);
            IO.println("Available: " + String.join(", ", allSnippets.keySet()));
            System.exit(1);
        }
        targets = List.of(allSnippets.get(key));
    } else {
        targets = allSnippets.values();
    }

    int count = 0;
    for (var s : targets) {
        var dir = Path.of(OUTPUT_DIR, s.category());
        Files.createDirectories(dir);
        var svg = generateSvg(s);
        Files.writeString(dir.resolve(s.slug() + ".svg"), svg);
        svgToPng(svg, dir.resolve(s.slug() + ".png"));
        count++;
    }
    IO.println("Generated %d SVG+PNG card(s) in %s/".formatted(count, OUTPUT_DIR));
}

File: html-generators/generateog.py
#!/usr/bin/env python3
"""
Generate Open Graph SVG+PNG cards (1200×630) for each pattern.
Light theme, side-by-side Old/Modern code, slug title at top.
Python equivalent of generateog.java — produces identical output.

Usage: python html-generators/generateog.py [category/slug]
       No arguments → generate all patterns.

Requires: cairosvg (pip install cairosvg)
"""

import json
import os
import re
import sys
import glob as glob_mod
from collections import OrderedDict

try:
    import yaml
except ImportError:
    yaml = None

try:
    import cairosvg
except ImportError:
    cairosvg = None

CONTENT_DIR = "content"
OUTPUT_DIR = "site/og"
CATEGORIES_FILE = "html-generators/categories.properties"

# ── Light-theme palette ─────────────────────────────────────────────────
BG         = "#ffffff"
BORDER     = "#d8d8e0"
TEXT       = "#1a1a2e"
TEXT_MUTED = "#6b7280"
OLD_BG     = "#fef2f2"
MODERN_BG  = "#eff6ff"
OLD_ACCENT = "#dc2626"
GREEN      = "#059669"
ACCENT     = "#6366f1"
BADGE_BG   = "#f3f4f6"

# ── Syntax highlight colors (VS Code light-inspired) ────────────────────
SYN_KEYWORD    = "#7c3aed"
SYN_TYPE       = "#0e7490"
SYN_STRING     = "#059669"
SYN_COMMENT    = "#6b7280"
SYN_ANNOTATION = "#b45309"
SYN_NUMBER     = "#c2410c"
SYN_DEFAULT    = "#1a1a2e"

JAVA_KEYWORDS = {
    "abstract", "assert", "boolean", "break", "byte", "case", "catch", "char",
    "class", "const", "continue", "default", "do", "double", "else", "enum",
    "extends", "final", "finally", "float", "for", "goto", "if", "implements",
    "import", "instanceof", "int", "interface", "long", "native", "new", "null",
    "package", "private", "protected", "public", "record", "return", "sealed",
    "short", "static", "strictfp", "super", "switch", "synchronized", "this",
    "throw", "throws", "transient", "try", "var", "void", "volatile", "when",
    "while", "with", "yield", "permits", "non-sealed", "module", "open", "opens",
    "requires", "exports", "provides", "to", "uses", "transitive",
    "true", "false",
}

SYN_PATTERN = re.compile(
    r"(?P<comment>//.*)|"
    r"(?P<blockcomment>/\*.*?\*/)|"
    r"(?P<annotation>@\w+)|"
    r'(?P<string>"""[\s\S]*?"""|"(?:[^"\\]|\\.)*"|\'(?:[^\'\\]|\\.)*\')|'
    r"(?P<number>\b\d[\d_.]*[dDfFlL]?\b)|"
    r"(?P<word>\b[A-Za-z_]\w*\b)|"
    r"(?P<other>[^\s])"
)

# ── Dimensions ──────────────────────────────────────────────────────────
W = 1200
H = 630
PAD = 40
HEADER_H = 100
FOOTER_H = 56
CODE_TOP = HEADER_H
CODE_H = H - HEADER_H - FOOTER_H
COL_W = (W - PAD * 2 - 20) // 2
CODE_PAD = 14
LABEL_H = 32
USABLE_W = COL_W - CODE_PAD * 2
USABLE_H = CODE_H - LABEL_H - CODE_PAD
CHAR_WIDTH_RATIO = 0.6
LINE_HEIGHT_RATIO = 1.55
MIN_CODE_FONT = 9
MAX_CODE_FONT = 16


# ── Helpers ─────────────────────────────────────────────────────────────

def load_properties(path):
    props = OrderedDict()
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            idx = line.find("=")
            if idx > 0:
                props[line[:idx].strip()] = line[idx + 1:].strip()
    return props


CATEGORY_DISPLAY = load_properties(CATEGORIES_FILE)


def read_auto(path):
    with open(path, encoding="utf-8") as f:
        if path.endswith((".yaml", ".yml")):
            if yaml is None:
                raise ImportError("PyYAML is required for YAML files: pip install pyyaml")
            return yaml.safe_load(f)
        return json.load(f)


def xml_escape(s):
    if s is None:
        return ""
    return (s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
             .replace('"', "&quot;").replace("'", "&apos;"))


def load_all_snippets():
    snippets = OrderedDict()
    for cat in CATEGORY_DISPLAY:
        cat_dir = os.path.join(CONTENT_DIR, cat)
        if not os.path.isdir(cat_dir):
            continue
        files = []
        for ext in ("json", "yaml", "yml"):
            files.extend(glob_mod.glob(os.path.join(cat_dir, f"*.{ext}")))
        files.sort()
        for path in files:
            data = read_auto(path)
            key = f"{data['category']}/{data['slug']}"
            snippets[key] = data
    return snippets


# ── Syntax highlighting ─────────────────────────────────────────────────

def highlight_line(line):
    if line == "...":
        return xml_escape(line)
    result = []
    last = 0
    for m in SYN_PATTERN.finditer(line):
        if m.start() > last:
            result.append(xml_escape(line[last:m.start()]))
        last = m.end()
        token = m.group()
        color = None
        if m.group("comment") or m.group("blockcomment"):
            color = SYN_COMMENT
        elif m.group("annotation"):
            color = SYN_ANNOTATION
        elif m.group("string"):
            color = SYN_STRING
        elif m.group("number"):
            color = SYN_NUMBER
        elif m.group("word"):
            if token in JAVA_KEYWORDS:
                color = SYN_KEYWORD
            elif token[0].isupper():
                color = SYN_TYPE
        if color:
            result.append(f'<tspan fill="{color}">{xml_escape(token)}</tspan>')
        else:
            result.append(xml_escape(token))
    if last < len(line):
        result.append(xml_escape(line[last:]))
    return "".join(result)


# ── SVG rendering ───────────────────────────────────────────────────────

def best_font_size(old_lines, modern_lines):
    max_chars = max(
        max((len(l) for l in old_lines), default=1),
        max((len(l) for l in modern_lines), default=1),
    )
    max_lines = max(len(old_lines), len(modern_lines))
    by_width = int(USABLE_W / (max_chars * CHAR_WIDTH_RATIO))
    by_height = int(USABLE_H / (max_lines * LINE_HEIGHT_RATIO))
    return max(MIN_CODE_FONT, min(MAX_CODE_FONT, min(by_width, by_height)))


def fit_lines(lines, font_size):
    line_h = int(font_size * LINE_HEIGHT_RATIO)
    max_lines = USABLE_H // line_h
    if len(lines) <= max_lines:
        return lines
    truncated = list(lines[:max_lines - 1])
    truncated.append("...")
    return truncated


def render_code_block(lines, x, y, line_h):
    parts = []
    for i, line in enumerate(lines):
        parts.append(
            f'    <text x="{x}" y="{y + i * line_h}" class="code" xml:space="preserve">'
            f'{highlight_line(line)}</text>\n'
        )
    return "".join(parts)


def generate_svg(data):
    left_x = PAD
    right_x = PAD + COL_W + 20
    label_y = CODE_TOP + 26
    code_y = CODE_TOP + 52

    old_lines = data["oldCode"].split("\n")
    modern_lines = data["modernCode"].split("\n")

    font_size = best_font_size(old_lines, modern_lines)
    line_h = int(font_size * LINE_HEIGHT_RATIO)

    old_lines = fit_lines(old_lines, font_size)
    modern_lines = fit_lines(modern_lines, font_size)

    cat_display = CATEGORY_DISPLAY.get(data["category"], data["category"])
    badge_width = len(cat_display) * 8 + 16

    old_code_svg = render_code_block(old_lines, left_x + 14, code_y, line_h)
    modern_code_svg = render_code_block(modern_lines, right_x + 14, code_y, line_h)

    return f"""<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="{W}" height="{H}" viewBox="0 0 {W} {H}">
  <defs>
    <style>
      .title    {{ font: 700 24px/1 'Inter', sans-serif; fill: {TEXT}; }}
      .category {{ font: 600 13px/1 'Inter', sans-serif; fill: {TEXT_MUTED}; }}
      .label    {{ font: 600 11px/1 'Inter', sans-serif; text-transform: uppercase; letter-spacing: 0.05em; }}
      .code     {{ font: 400 {font_size}px/1 'JetBrains Mono', monospace; fill: {TEXT}; }}
      .footer   {{ font: 500 13px/1 'Inter', sans-serif; fill: {TEXT_MUTED}; }}
      .brand    {{ font: 700 14px/1 'Inter', sans-serif; fill: {ACCENT}; }}
    </style>
    <clipPath id="clip-left">
      <rect x="{left_x}" y="{CODE_TOP}" width="{COL_W}" height="{CODE_H}" rx="8"/>
    </clipPath>
    <clipPath id="clip-right">
      <rect x="{right_x}" y="{CODE_TOP}" width="{COL_W}" height="{CODE_H}" rx="8"/>
    </clipPath>
  </defs>

  <!-- Background -->
  <rect width="{W}" height="{H}" rx="16" fill="{BG}"/>
  <rect x="0.5" y="0.5" width="{W - 1}" height="{H - 1}" rx="16" fill="none" stroke="{BORDER}" stroke-width="1"/>

  <!-- Header: category badge + title -->
  <rect x="{PAD}" y="28" width="{badge_width}" height="22" rx="6" fill="{BADGE_BG}"/>
  <text x="{PAD + 8}" y="43" class="category">{xml_escape(cat_display)}</text>
  <text x="{PAD}" y="76" class="title">{xml_escape(data['title'])}</text>

  <!-- Left panel: Old code -->
  <rect x="{left_x}" y="{CODE_TOP}" width="{COL_W}" height="{CODE_H}" rx="8" fill="{OLD_BG}"/>
  <rect x="{left_x}" y="{CODE_TOP}" width="{COL_W}" height="{CODE_H}" rx="8" fill="none" stroke="{BORDER}" stroke-width="0.5"/>
  <text x="{left_x + 14}" y="{label_y}" class="label" fill="{OLD_ACCENT}">\u2717  {xml_escape(data['oldLabel'])}</text>
  <g clip-path="url(#clip-left)">
{old_code_svg}  </g>

  <!-- Right panel: Modern code -->
  <rect x="{right_x}" y="{CODE_TOP}" width="{COL_W}" height="{CODE_H}" rx="8" fill="{MODERN_BG}"/>
  <rect x="{right_x}" y="{CODE_TOP}" width="{COL_W}" height="{CODE_H}" rx="8" fill="none" stroke="{BORDER}" stroke-width="0.5"/>
  <text x="{right_x + 14}" y="{label_y}" class="label" fill="{GREEN}">\u2713  {xml_escape(data['modernLabel'])}</text>
  <g clip-path="url(#clip-right)">
{modern_code_svg}  </g>

  <!-- Footer -->
  <text x="{PAD}" y="{H - 22}" class="footer">JDK {data['jdkVersion']}+</text>
  <text x="{W - PAD}" y="{H - 22}" class="brand" text-anchor="end">javaevolved.github.io</text>
</svg>
"""


def svg_to_png(svg_content, png_path):
    if cairosvg is None:
        raise ImportError("cairosvg is required for PNG generation: pip install cairosvg")
    cairosvg.svg2png(
        bytestring=svg_content.encode("utf-8"),
        write_to=png_path,
        output_width=W * 2,
        output_height=H * 2,
    )


# ── Main ────────────────────────────────────────────────────────────────

def main():
    all_snippets = load_all_snippets()
    print(f"Loaded {len(all_snippets)} snippets")

    # Filter to a single slug if provided
    if len(sys.argv) > 1:
        key = sys.argv[1]
        if key not in all_snippets:
            print(f"Unknown pattern: {key}")
            print(f"Available: {', '.join(all_snippets.keys())}")
            sys.exit(1)
        targets = {key: all_snippets[key]}
    else:
        targets = all_snippets

    count = 0
    for key, data in targets.items():
        cat = data["category"]
        slug = data["slug"]
        out_dir = os.path.join(OUTPUT_DIR, cat)
        os.makedirs(out_dir, exist_ok=True)

        svg = generate_svg(data)
        svg_path = os.path.join(out_dir, f"{slug}.svg")
        with open(svg_path, "w", encoding="utf-8") as f:
            f.write(svg)

        png_path = os.path.join(out_dir, f"{slug}.png")
        svg_to_png(svg, png_path)
        count += 1

    print(f"Generated {count} SVG+PNG card(s) in {OUTPUT_DIR}/")


if __name__ == "__main__":
    main()
